src/compiler/optim.jl

```diff
@@ -1,28 +1,73 @@
-    tm = AMDGPUnative.machine(ctx.agent, triple(mod))
+function optimize!(job::CompilerJob, mod::LLVM.Module, entry::LLVM.Function)
+    tm = AMDGPUnative.machine(job.agent, triple(mod))
 
-    if ctx.kernel
-        entry = promote_kernel!(ctx, mod, entry)
+    if job.kernel
+        entry = promote_kernel!(job, mod, entry)
     end
 
-    let pm = ModulePassManager()
-        global global_ctx
-        global_ctx = ctx
-
+    function initialize!(pm)
         add_library_info!(pm, triple(mod))
         add_transform_info!(pm, tm)
-        internalize!(pm, [LLVM.name(entry)])
+    end
+
+    global current_job
+    current_job = job
+
+    # Julia-specific optimizations
+    #
+    # NOTE: we need to use multiple distinct pass managers to force pass ordering;
+    #       intrinsics should never get lowered before Julia has optimized them.
+    if VERSION < v"1.2.0-DEV.375"
+        # with older versions of Julia, intrinsics are lowered unconditionally so we need to
+        # replace them with GPU-compatible counterparts before anything else. that breaks
+        # certain optimizations though: https://github.com/JuliaGPU/CUDAnative.jl/issues/340
+
+        ModulePassManager() do pm
+            initialize!(pm)
+            add!(pm, FunctionPass("LowerGCFrame", lower_gc_frame!))
+            aggressive_dce!(pm) # remove dead uses of ptls
+            add!(pm, ModulePass("LowerPTLS", lower_ptls!))
+            run!(pm, mod)
+        end
 
-        # lower intrinsics
-        add!(pm, FunctionPass("LowerGCFrame", lower_gc_frame!))
-        aggressive_dce!(pm) # remove dead uses of ptls
-        add!(pm, ModulePass("LowerPTLS", lower_ptls!))
+        ModulePassManager() do pm
+            initialize!(pm)
+            ccall(:jl_add_optimization_passes, Cvoid,
+                  (LLVM.API.LLVMPassManagerRef, Cint),
+                  LLVM.ref(pm), Base.JLOptions().opt_level)
+            run!(pm, mod)
+        end
+    else
+        ModulePassManager() do pm
+            initialize!(pm)
+            ccall(:jl_add_optimization_passes, Cvoid,
+                  (LLVM.API.LLVMPassManagerRef, Cint, Cint),
+                  LLVM.ref(pm), Base.JLOptions().opt_level, #=lower_intrinsics=# 0)
+            run!(pm, mod)
+        end
 
-        ccall(:jl_add_optimization_passes, Cvoid,
-              (LLVM.API.LLVMPassManagerRef, Cint),
-              LLVM.ref(pm), Base.JLOptions().opt_level)
+        ModulePassManager() do pm
+            initialize!(pm)
+
+            # lower intrinsics
+            #add!(pm, FunctionPass("LowerGCFrame", lower_gc_frame!))
+            aggressive_dce!(pm) # remove dead uses of ptls
+            add!(pm, ModulePass("LowerPTLS", lower_ptls!))
+
+            # the Julia GC lowering pass also has some clean-up that is required
+            if VERSION >= v"1.2.0-DEV.531"
+                late_lower_gc_frame!(pm)
+            end
+
+            run!(pm, mod)
+        end
+    end
+
+    # PTX-specific optimizations
```

Comment:
@vchuravy do you think these passes could be useful for us? Or are they just wasted compilation time?